JAMES-2808 serialization/deserialization DeletedMessageWithStorageInformation

Arsnael opened the pull request:
Not finished...

Not sure I'm on the right path : 
* I feel like I should refactor first commit to be more like second commit, I think I should have custom serializers for `BucketName` and `BlobId`
* I'm stuck on my second commit, it seems my test cannot deserialize correctly the `recipients` : 
```
java.lang.IllegalArgumentException: No serializer found for class org.apache.james.core.Domain and no properties discovered to create BeanSerializer (to avoid exception, disable SerializationFeature.FAIL_ON_EMPTY_BEANS) (through reference chain: org.apache.james.vault.DeletedMessage["recipients"]->com.google.common.collect.RegularImmutableList[0]->org.apache.james.core.MailAddress["domain"])
```
I do have a custom deserializer on `Domain` though... I don't really get what is wrong...

chibenwa commented on mailbox/plugin/deleted-messages-vault/pom.xml:
```diff
@@ -84,10 +84,23 @@
         </dependency>
+        <dependency>
+            <groupId>${james.groupId}</groupId>
+            <artifactId>james-server-jmap</artifactId>
```
Why do we need this dependency?

chibenwa commented on mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/DeletedMessage.java:
```diff
@@ -169,9 +171,17 @@ public DeletedMessage build() {
-                          ZonedDateTime deliveryDate, ZonedDateTime deletionDate, MaybeSender sender, List<MailAddress> recipients,
-                          Optional<String> subject, boolean hasAttachment, long size) {
+    @JsonCreator
+    public DeletedMessage(@JsonProperty("messageId") MessageId messageId,
```
We should have a DTO

chibenwa commented on mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/json/ObjectMapperVaultFactory.java:
```diff
@@ -0,0 +1,161 @@
+        coreModule.addDeserializer(MaybeSender.class, new MaybeSenderDeserializer());
+        coreModule.addSerializer(MaybeSender.class, new MaybeSenderSerializer());
+        coreModule.addDeserializer(User.class, new UserDeserializer());
+        coreModule.addSerializer(User.class, new UserSerializer());
```
With a DTO all these serializer / deserializers wouldn't be needed.

chibenwa commented on server/blob/blob-api/pom.xml:
```diff
@@ -51,6 +51,10 @@
         </dependency>
+        <dependency>
+            <groupId>com.fasterxml.jackson.core</groupId>
+            <artifactId>jackson-annotations</artifactId>
```
Why ? Do I really need jackson to use the BlobStore ?

chibenwa commented on mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/metadata/StorageInformation.java:
```diff
@@ -20,13 +20,46 @@
 
 public class StorageInformation {
+
+    public static class Builder {
```
Let's do a staged builder here then

Arsnael replied:
Oups forgot about that one, will fix

Arsnael commented:
> DOn't forget POJO tests & JSON tests

Yep I know. WIP I just wanted to present my work at its current state. Will take care of the tests today !

Arsnael commented:
> I dont see the serializer

Not sure to get your point? What I can constat is that usually you use the DTO with `JsonTransformer` and `JsonExtractor` classes to serialize and deserialize... and I have tests with an `ObjectMapper` to simulate just that. Or did I misunderstand something?

chibenwa commented on mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/dto/DeletedMessageWithStorageInformationDTO.java:
```diff
@@ -0,0 +1,272 @@
+            return size;
+        }
+
+        private static ImmutableList<String> getOriginMailboxes(List<MailboxId> originMailboxes) {
```
static goes first

chibenwa replied:
serializeOriginMailboxes

Arsnael replied:
I don't think static goes before the private..?

chibenwa replied:
static fields

public static

private static

fields

constructor

public methods

private methods

Arsnael replied:
Misunderstood, sorry

chibenwa commented on mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/metadata/StorageInformation.java:
```diff
@@ -27,6 +27,23 @@
+    }
+
+    public static Builder.RequireBucketName builder() {
+        return bucketName -> blobId -> new StorageInformation(bucketName, blobId);
```
The class constructor `StorageInformation` then needs to be private

Arsnael commented:
```
[c4ee15a29fd0b400c33bdc32be5a3c7a2e5bfc03] [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.7.0:compile (default-compile) on project apache-james-mailbox-deleted-messages-vault: Compilation failure
[c4ee15a29fd0b400c33bdc32be5a3c7a2e5bfc03] [ERROR] /james-project/mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/blob/BlobStoreDeletedMessageVault.java:[61,28] StorageInformation(org.apache.james.blob.api.BucketName,org.apache.james.blob.api.BlobId) has private access in org.apache.james.vault.metadata.StorageInformation
```
Rebased on master and fixed the issue

chibenwa commented:
```
[3a56b5bfff1ce2897e1ae9c4f0488cdd13cf73d3] [ERROR] enqueueShouldNotFailWhenBrowsingIterating  Time elapsed: 0.025 s  <<< ERROR!
[3a56b5bfff1ce2897e1ae9c4f0488cdd13cf73d3] java.util.NoSuchElementException
[3a56b5bfff1ce2897e1ae9c4f0488cdd13cf73d3] 	at org.apache.james.queue.api.ManageableMailQueueContract.enqueueShouldNotFailWhenBrowsingIterating(ManageableMailQueueContract.java:296)
[3a56b5bfff1ce2897e1ae9c4f0488cdd13cf73d3] 
[3a56b5bfff1ce2897e1ae9c4f0488cdd13cf73d3] [INFO] 
[3a56b5bfff1ce2897e1ae9c4f0488cdd13cf73d3] [INFO] Results:
[3a56b5bfff1ce2897e1ae9c4f0488cdd13cf73d3] [INFO] 
[3a56b5bfff1ce2897e1ae9c4f0488cdd13cf73d3] [ERROR] Errors: 
[3a56b5bfff1ce2897e1ae9c4f0488cdd13cf73d3] [ERROR]   ManageableMailQueueContract.enqueueShouldNotFailWhenBrowsingIterating:296 » NoSuchElement
```

test this please

mbaechler commented on mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/dto/DeletedMessageWithStorageInformationConverter.java:
```diff
@@ -0,0 +1,94 @@
+        this.mailboxIdFactory = mailboxIdFactory;
+    }
+
+    public StorageInformation toStorageInformation(DeletedMessageWithStorageInformationDTO.StorageInformationDTO storageInformationDTO) {
```
both input and output are actually `StorageInformation`
I would call methods `toDomainObject` and `toDTO` and use method overloading for handling different types

mbaechler commented on mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/dto/DeletedMessageWithStorageInformationDTO.java:
```diff
@@ -0,0 +1,272 @@
+        }
+
+        @Override
+        public final boolean equals(Object o) {
```
do you need equals/hashcode for a DTO?

Arsnael replied:
Needed it for the test, to compare the DTOs. Is that wrong?

chibenwa replied:
> Needed it for the tests, to compare the DTOs. Is that wrong?

We do compare domain object, we do compare JSON generated from DTOs, there's no reason to compare the DTOs directly?

mbaechler commented on mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/dto/DeletedMessageWithStorageInformationDTO.java:
```diff
@@ -0,0 +1,272 @@
+                serializeOriginMailboxes(deletedMessage.getOriginMailboxes()),
+                deletedMessage.getOwner().asString(),
+                deletedMessage.getDeliveryDate().toString(),
+                deletedMessage.getDeletionDate().toString(),
```
If you can, avoid using toString for anything else than debug

chibenwa commented:
`SpamAssassinContract.expungingSpamMessageShouldNotImpactSpamAssassinState`

test this please

blackheaven commented on mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/dto/DeletedMessageWithStorageInformationConverter.java:
```diff
@@ -0,0 +1,94 @@
+
+    private ImmutableList<MailAddress> deserializeRecipients(List<String> recipients) {
+        return recipients.stream()
+            .map(Throwing.function(recipient -> new MailAddress(recipient)))
```
sneakyThrow?

chibenwa replied:
What would it bring?

blackheaven replied:
we do it everywhere else, why making an exception here?

chibenwa replied:
Well because explicit exception handling is verbose while throwing runtime is not.

What's important is: do we get notified when something goes wrong? (yes in both case)

Do we abort a complex computation for a single error? (That's the responsibility from my Cassandra code after all, and explicit exception handling would force me to handle it)

Arsnael commented:
Tests are :green_apple: , failed at deployment tests : 
```
[fb740aee6961ae009da2214229c9bce6ff48195c] [ERROR] Errors: 
[fb740aee6961ae009da2214229c9bce6ff48195c] [ERROR]   DeploymentValidation.validateDeployment:81 » InvalidServerResponse 
[fb740aee6961ae009da2214229c9bce6ff48195c] Location: ...
[fb740aee6961ae009da2214229c9bce6ff48195c] [ERROR]   DeploymentValidation.validateDeploymentWithMailsFromSmtp:87 » SMTPSending Erro...
```
test this please

Arsnael commented:
I approve @chibenwa fixup. Thanks!

chibenwa commented:
Merged
